### DISCUSSION TITLE
Handling different origin heights

### DISCUSSION THREAD

**Discussion by paulmelis** [StereoKit/StereoKit#343] [Q&A]
Do you have a best practice on how to handle the different origins for the same project on different systems/headsets? I noticed that with the XR-3 using SteamVR tracking the world's Y=0 is at floor level, while with a Hololens it is the physical headset height at application start, and similar when starting a project on a Quest through Link it seems.

I actually like the Steam VR origin height best, as it gives the floor level as reference height, which isn't easily available on a HL2 automatically (I suspect the Quest might be able to get it from it's spatial map?). Plus the SteamVR origin stays in the same place over different program runs, while for the other devices it is based on startup position, but that's a different aspect. Having a known floor Y value makes it easier to position stuff relative to that, although sometimes you also want to position something X distance below a user's head.

Anyway, I figured I might be able to detect the current device and then compute a floor Y value based on an estimate of the user's height, in case the device does not use Steam VR tracking. But I don't actually see any device-related info in the SK API, unless I'm missing it (I looked in `SK.System`)?

PS Sorry for the barrage of questions/issues lately, I've just been spending quite a bit of time developing with SK recently, which generates those things :smile:

**Comment by maluoi** [comment]
StereoKit's origin currently _should_ always be where the headset was located during startup, and this is related largely to OpenXR's reference space concept!

## Reference Spaces

The main important types of reference spaces are:
- Local (what SK does)
- Stage (think VR playspaces with guardian boundaries)
- There's a [few additional variations](https://www.khronos.org/registry/OpenXR/specs/1.0/html/xrspec.html#reference-spaces) of these, but not particularly important.

_All_ runtimes support Local space, which is why StereoKit uses it, and only some provide a Stage! StereoKit does provide information about the Stage space if present, via `World.HasBounds`, `World.BoundsPose`, and `World.BoundsSize`.

I'd agree that Stage would be a much better default, or at least an option to use as the main reference space when present! If I did go this route though, I'd want to provide a reasonably reliable fallback for emulating Stage when not present, but this is unfortunately not exactly trivial. I'm still holding out some hope for some sort of extension for floor height! This is stuff I'll probably revisit and re-evaluate when I get to a breaking change version, but with the `World` class, I'm not too concerned about it in the near future.

## Best Practices

For StereoKit demos, I will typically use `World.HasBounds ? World.BoundsPose.position.y : -1.5f` for the floor height. This is a decent fast approximate, but a robust application might want a little more! It generally assumes the user is standing, and is roughly 1.5m tall :D 

On HoloLens, you could probably use some smart `World.Raycast`ing to find a floor height, but you'd have to watch for chairs and tables! Scene Understanding can technically find labeled "floor" planes as well, but StereoKit doesn't surface this information just yet.

However, _**the most reliable method**_ is generally to build a menu for it, and just ask the user about their height! Many consumer grade applications will do this _**even a Stage is present**_! This is handy from an accessibility standpoint too, and can allow sitting users, or short users to fake a height that's better for them, but may not match the room's attributes.

It can also be good to note that the Stage may change while the application is running, if the user adjusts their guardian/boundary setup.

---

And no worries about the questions, thanks for asking them! It's helpful for me too! Also, this stuff is all durable + searchable, so others will totally benefit from it as well :)

**Reply by paulmelis** [reply]
Thanks for the details. I will test some more, but just found that running a Quest through Link has `World.HasBounds` set to `true` and `World.BoundsPose.position.y` at 0.00. Which is a bit unexpected given what you wrote above. 

For now I've used a dirty hack to check (each frame) the y position of the both hand palms and use that to keep track of the lowest y value of the floor encountered, which is then used during rendering as y translation :smile: But it'd really like to avoid this trick and only use it in this specific configuration (Quest+Link). Is there any way to get the device name, something like `systemName` through SK? Or should I query OpenXR directly?

**Reply by maluoi** [reply]
There is actually a bug that popped up specifically in v0.3.6-preview.4 that is likely what you're seeing here! (I believe that's what you're on right now?) I've fixed it, but haven't pushed up a new build with that fix yet.

Not 100% sure, but if you're doing standing room only bounds for Quest, try it out with room bounds instead? It _might_ get you past the bug until I push the new stuff up.

**Reply by paulmelis** [reply]
No, I'm on 0.3.5 still. I'll do some more tests later (after dinner :wink:). This was indeed with stationary guardian, maybe that has to do with it
